### ISSUE TITLE
Загрузчик 7625

### ISSUE THREAD

**Issue by snegin** [forth32/qtools#10]
Пожалуйста сделайте патченный загрузчик под этот проц.

**Comment by forth32** [comment]
Это в смысле MSM7625? А непатченный загрузчик под этот чипсет есть?

**Comment by snegin** [comment]
Да,MSM7626A.Есть.

**Comment by snegin** [comment]
https://yadi.sk/d/G-lEpkEp3GCo6L

**Comment by snegin** [comment]
Подойдет этот загрузчик?

**Comment by forth32** [comment]
Кто ж его знает. У меня нет ни одного устройства на этом чипсете, как мне проверить?
Кстати, а какие вообще распространенные устройства на нем есть?

**Comment by vvevvevve** [comment]
На этих чипсетах делали ещё и устройства с eMMC в качестве носителя.
**snegin**: 
1. Ваш аппарат точно с NAND? Как он называется?
2. В каком режиме находится Ваш аппарат? Что доступно - интерфейс PBL, интерфейс SBL?
3. Загрузчик, который Вы выложили - без заголовка. Куда делся заголовок? В оригинале, это был NPRG, ENPRG или MPRG? Откуда Вы, вообще, взяли этот код?
4. Вы пробовали использовать наш "комплектный" ENPRG7x27Ap.bin ?

**Comment by snegin** [comment]
to forth32
Я в том плане можно ли пропатчить этот загрузчик.Понятно что без устройства не протестируешь.
Novatel 4082
Huawei E585

to vvevvevve
1.С нандом.Сегодня разберу и посмотрю точно какая там память.
2.Аппарат  полностью в рабочем режиме.Есть 2 стандартных порта.
3.Загрузчик вырезан из софта для анлока.Судьба заголовка мне неизвестна.Думаю что изначально это был NPRG.
4.Не провобовал.Заведу в PBL - попробую.

Мне нужно прочитать efs и слить разделы устройства.

**Comment by vvevvevve** [comment]
Вот, нашёл какое-то чудо: http://vve.su/files/misc/tmp/NPRG7627A.bin .
Попробуйте просто запустить его в исходном виде, непатченный. Думаю, что это нужно делать так:
`qdload -pX -k9 -q -a2dc00000 NPRG7627A.bin`
Если он запустится и успешно выполнится команда hello, то пропатчим его и включим в комплект.

**Comment by vvevvevve** [comment]
Да, X - это номер порта интерфейса SBL, после переключения аппарата в download mode.

**Comment by snegin** [comment]
Ребут после отправки лоадера. Думаю что 7627 и 7625 разные.В 7627 есть пбл.У меня получалось их заводить в него.В 7625 ничего не получилось.

**Comment by snegin** [comment]
Кажется появилась мысль где загрузчик можно взять.Есть один апдейтер. В понедельник выложу.

**Comment by vvevvevve** [comment]
Они все разные, в той или иной мере. И PBL есть во всех 7xxx. И ОЗУ на 7x25 может быть размещено в диапазоне от 0 до 40000000, так что 2dc00000 "вписывается" в этот диапазон.
А может, этот ребут возникает на этапе попытки выполнения hello? Попробуйте запускать без ключа -q.

**Comment by vvevvevve** [comment]
Да, и базовый адрес NANDC в 7x25, абсолютно достоверно, равен a0a00000, как у нас и предполагается.

**Comment by snegin** [comment]
hello ни при чём.Сразу после отправки загрузчика перезагрузка.В Вашем загрузчике нет NAND,которая стоит в роутере.Я думаю что хелло не даст ответа при отсутствии в загрузчике этой NAND.Хотя перезагружаться он не должен.Встречал я "творения" самсунга и хуавеи у которых встречалось аналогичное.При этом PBL загрузчики на них работали отлично,а вот SPL ребут и все...Но не на всех прошивках - на некоторых все работало. Апдейтеры их шили и конечно при SBL. Как вариант поищу апдейтер  под эту модель.

**Comment by snegin** [comment]
HYNIX_H8BCS0PHOMCP-56M - NAND.

**Comment by snegin** [comment]
https://yadi.sk/d/u3JK6Sf53GMaRQ
Апдейтер в котором должен быть рабочий загрузчик.При распаковке он скорее всего лежит в файле 505 вместе с прошивкой,но необходимо найти начало и конец.С этим проблемы.

**Comment by carver-ua** [comment]
это же хуавей, в ресурсах обычный их app.
сам appраспковывается скрипом от 2009 года,
split_updata.pl, #   Author(s)     : McSpoon, #   Description   : Unpack a Huawei U8220 'UPDATA.APP' file.
никаких бутов там нет, они в хуавея не используются.

**Comment by snegin** [comment]
Понятно.С хуавеями не знаком.Есть и другие загрузчики под 7627.Одна надежда на PBL.А он там точно есть?

**Comment by snegin** [comment]
https://yadi.sk/d/4diVywer3GMcND
Ещё загрузчики.

**Comment by vvevvevve** [comment]
> hello ни при чём.Сразу после отправки загрузчика перезагрузка.
Если запускаете qdload с ключом -q или -i, то hello отправляется автоматически. После запуска загрузчика, конечно. Но разумнее в таких случаях запускать утилиту без этих ключей.

> В Вашем загрузчике нет NAND,которая стоит в роутере.Я думаю что хелло не даст ответа при отсутствии в загрузчике этой NAND
Даст ответ. "Unrecognized flash" или вроде того.

> Есть и другие загрузчики под 7627
Ну так пробуйте их запускать, проверяйте. Чего Вы ждёте?

> Одна надежда на PBL.А он там точно есть?
Да какая разница? Если он даже не называется PBL, всё равно в ПЗУ платформы есть код, который умеет инициализировать железо, загружать и запускать ENPRG.

**Comment by vvevvevve** [comment]
> Ещё загрузчики.

Ну и пробуйте их запускать. Смотрите адреса в заголовках (они там у Вас все разные) и задавайте в ключе -a: (слово по смещению 0xc в файле) - 0x28.

**Comment by snegin** [comment]
Сейчас попробую.

**Comment by vvevvevve** [comment]
Кстати, тот ENPRG7x27Ap.bin в комплекте - это, по идентификатору образа, NPRG, а не ENPRG. Так что его можно тоже пробовать грузить (по адресу 600000).

**Comment by snegin** [comment]
Загрузчик отправлен.Дальше отсылка hello:
Отсылка hello... повторный hello возвратил ошибку!
00000000: 7e 03 00 03 33 1b 7e                             *~...3.~         *
Ребут.
Это о ENPRG7x27Ap.bin.

**Comment by vvevvevve** [comment]
Видите, он даже запускается и что-то успевает ответить.
Запускайте qdload без hello, а hello потом пробуйте отправить в интерактивном режиме qcommand, командой is .

**Comment by vvevvevve** [comment]
Или хотя бы просто без параметров - посмотреть, что ответит: команда **c 0** в интерактивном режиме.

**Comment by carver-ua** [comment]
по коду ошибки,  похоже что после отправки загрузчика, ему забыли передать управление.
поэтому телефон на 01 и отвечает 03 с кодом 0003 - неправильная длина. 
(NAK response packet+Invalid length)
потому что 01 это еще не hello a пока white.

**Comment by snegin** [comment]
---- ответ ---
00000000: 7e 03 00 06 9e 4c 7e                             *~....L~         *
Что я делаю не так?

**Comment by snegin** [comment]
ДМ - qdload -p21 -k9 -a600000 loaders\ENPRG7x27Ap.bin

**Comment by snegin** [comment]
С остальными перезагрузка.

**Comment by vvevvevve** [comment]
> Что я делаю не так?

Я ведь вижу только ответ, но не вижу - что и на что отвечает. Покажите полный сеанс загрузки NPRG и выполнения команды.

**Comment by carver-ua** [comment]
попробуйте просто профлешить другое похожее устройство, и снифером посмотреть как все  устроенно на квалкоме. потом сравнить оба лога, и все станет намного понятнее. пока выходит что у вас все буты не те. или в ребут или не стартуют. может вы не туда грузите, или не то. 
может есть тема на 4pda, или еще где, по вашему устройству на 7625, или не 7625.
думаю на многих форумах - есть ребята, которые по удаленке - и бут нужный найдут, и допилят его, и помогут вам поднять устройство.

**Comment by vvevvevve** [comment]
> потому что 01 это еще не hello a пока white.

Это в DMSS команда 01 = Write, а в SDP это Hello. Он же не SBL эту команду отправляет, а загрузчику. Я надеюсь, во всяком случае... Почему и прошу показать полную картину сессии.

**Comment by carver-ua** [comment]
ARMPRG не может ответить 03, там ответы четные восновном.
значит OEMSBL не стартонул ARMPRG, и отвечает сам. 
ARMPRG кривой или адресс, х.з. 
может канешно ответ на 05/GO - чем-то поможет, 
но в случае перегрузки - он не поможет, а там где не стартует, явно дело в буте.

подозреваю производитель устройства, версии с DIAG/ATI, дамп ram, и еще что,  было бы намного информативнее чем фраз бут для 7625.

**Comment by snegin** [comment]
https://yadi.sk/d/Bt4Bog0u3GMrMt
 Слил купстом.Может чем-то поможет.

**Comment by carver-ua** [comment]
да, там сразу в cmm написанно что Huawei E5832
7625A-13553008-SNCAALZM

**Comment by carver-ua** [comment]
лет пять назад, у меня был его близкий родственник, huawei ec5825, в oemsbl
7625A-135520-SNCABLUM

я тогда нашел пару NPRG7625 в гугле, от acer e101 и wellcom a66
но вроде у меня тогда ничего не получилось, но уже не помю с чем.
толи с адресом загрузки, толи с дописыванием в бут своего флешера. 
адресс загрузки в тех бутах был 0x05F00000, а qtools тогда ее небыло.

тут не помню, но на старых платформах адрес мог быть разный, 
иногда нужно было сделать "rebase" чужому загрузчику, 
ну поправить скриптом все смещения в файле.

**Comment by carver-ua** [comment]
вот линк на те буты. может там только дрова придется самому допилить. 
https://www.4shared.com/rar/T_jjXg4Qei/nprg7625.html
acer e101 0502_9002 NPRG7625.hex
acer e101 05C6_9002 NPRG7625.hex
wellcom a66 AQ4-0-0360-NPRG7625.hex
wellcom a66 AQ4-0-0370-NPRG7625.hex

**Comment by snegin** [comment]
В любом случае огромное спасибо.О результатах напишу.

**Comment by carver-ua** [comment]
судя по 80-VP758, в msm7225(думаю всех 7x25 тоже) - отсутствует PBL. тоесть загрузчик в процессоре. 
и вообще по большинству пунктов *Cannot be supported, due to hardware issue.

тоесть, в отличии от новых процов, qtools на данном устройстве, пригодится в очень редких ситуациях, 
скорее в одной наверное, переписать/поправить OEMINFO, на рабочем устройстве. 

востановить с довндоада - скорее всего можно стандартным флешером, 
а если полностью труп, то только jtag, а там пиши что угодно.

**Comment by vvevvevve** [comment]
> отсутствует PBL. тоесть загрузчик в процессоре

Да не отсутствует он - как же без него? Просто из-за какой-то ошибки у них не получилось организовать загрузку кода. Даже не могу представить - какая это могла бы быть ошибка, ведь сама по себе и флешка пишется, и порт USB работает. А одновременно эти действия сразу после запуска почему-то не получается. 
То есть, да: особого смысла что-то читать из него нет. Разве только для того, чтобы потом c JTAG-ом записать.

**Comment by carver-ua** [comment]
наверное первой фразой я должен написать спасибо, 
и вам/тебе, и forth32, за помощь в других ситуациях, не связанных с qtools.

у меня нет большого опыта, но PBL в квалкомах появился "недавно".
когда  другие процы уже грузились с sd/usb/232, может что забыл, 
а тот-же i7 - наверняка до сих пор биос просит.

на этой пасочке, думаю это просто квалкомовский download, 
после dmss комманды 3a или проблем с amss(на lte - dsp2)

нет там PBL, и armprg, неважно какие, все-равно пока левак.

просто топикстартер не в тот саппорт написал, 
квалкому нужно было писать, а на на github )))

**Comment by newadb** [comment]
1. У меня аналогичная проблема (mf677).
qdload -p49 -k1 -i -t -d50 ENPRG8200p.bin

 Найден блок идентификации загрузчика
 Не указан тип чипсета - полная инициализация невозможна

 Файл загрузчика: ENPRG8200p.bin
 Адрес загрузки: 58020000
 Размер файла: 162084
 Загружено: 162084
 Запуск загрузчика...ok
 Отсылка hello... повторный hello возвратил ошибку!
00000000: 15 01 51 43 4f 4d 20 66 61 73 74 20 64 6f 77 6e  *..QCOM fast down*
00000010: 6c a8 8b 7e                                      *l..~

2. Ну и гораздо более тривиальная:
![image](https://user-images.githubusercontent.com/31523931/29949297-738a8106-8ebc-11e7-92bc-3eeed92232c4.png)
